### PR TITLE
DiagGGN: Extended to convnd & convtransposend

### PR DESCRIPTION
**Pull request by sbharadwajj**
Note: It is better to review this only after [this pull request](https://github.com/f-dangel/backpack/pull/112) and [this pull request](https://github.com/f-dangel/backpack/pull/114) is merged. 

Because we are changing the function in `utils/conv.py` and `utils/conv_transpose.py`, so consecutive extensions have to be changed.

Note: the file `diag_hessian/conv2d.py` is changed only so that the tests pass

Extension to `convnd` and `convtransposend`

### CODE REVIEW

File: backpack/utils/conv.py [hunk 1]
```diff
@@ -49,23 +49,37 @@ def separate_channels_and_pixels(module, tensor):
...
     """
     V_axis, N_axis = 0, 1
-    bias_diagonal = (einsum("vnchw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+
+    if N == 1:
+        bias_diagonal = (einsum("vncl->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 2:
+        bias_diagonal = (einsum("vnchw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 3:
+        bias_diagonal = (einsum("vncdhw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
```
Comment (f-dangel) [inline backpack/utils/conv.py:82]: The `**2.sum([...])` part is identical for all three. Can you pull that out and instead create the correct equation string for `einsum` in the `if` statement?

Comment (f-dangel) [reply backpack/utils/conv.py:82]: Can you also raise a `ValueError` if `N` is not `1,2,3`?

File: backpack/utils/conv_transpose.py [hunk 2]
```diff
@@ -23,6 +24,53 @@ def get_weight_gradient_factors(input, grad_out, module, N):
...
+    `sqrt` must be the backpropagated quantity for DiagH or DiagGGN(MC)
+    """
+    V_axis, N_axis = 0, 1
+
+    if N == 1:
+        bias_diagonal = (einsum("vncl->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 2:
+        bias_diagonal = (einsum("vnchw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
+    elif N == 3:
+        bias_diagonal = (einsum("vncdhw->vnc", sqrt) ** 2).sum([V_axis, N_axis])
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:70]: Can you pull out the `einsum, **2, .sum([...])` to reduce duplication? Also, raise a `ValueError` if `N` is not `1,2,3`

File: backpack/extensions/secondorder/diag_hessian/conv2d.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def bias(self, ext, module, g_inp, g_out, backproped):
         h_diag = torch.zeros_like(module.bias)
 
         for h_sqrt, sign in zip(sqrt_h_outs, sqrt_h_outs_signs):
-            h_diag_curr = convUtils.extract_bias_diagonal(module, h_sqrt)
+            h_diag_curr = convUtils.extract_bias_diagonal(module, h_sqrt, N=2)
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_hessian/conv2d.py:19]: @sbharadwajj can you confirm that this is just a modification because the interface of `extract_{bias,weight}_diagonal` changes due to the support for `ConvNd` in `DiagGGN{MC,Exact}`?

File: backpack/utils/conv.py [hunk 4]
```diff
@@ -49,23 +49,37 @@ def separate_channels_and_pixels(module, tensor):
     return eingroup("v,n,c,h,w->v,n,c,hw", tensor)
 
 
-def extract_weight_diagonal(module, input, grad_output):
+def extract_weight_diagonal(module, input, grad_output, N):
     """
     input must be the unfolded input to the convolution (see unfold_func)
     and grad_output the backpropagated gradient
     """
-    grad_output_viewed = separate_channels_and_pixels(module, grad_output)
```
Comment (f-dangel) [inline backpack/utils/conv.py:57]: Is this `separate_channels_and_pixels`function used somewhere else? If so, why not just generalize it to be able to handle `N=1,2,3`? If not, then deprecate that function

File: test/extensions/secondorder/diag_ggn/diaggnn_settings.py [hunk 5]
```diff
@@ -4,11 +4,238 @@
...
+        "module_fn": lambda: torch.nn.Sequential(
+            torch.nn.ConvTranspose3d(3, 2, 2, padding=2, dilation=2, stride=2),
+            torch.nn.ReLU(),
+            torch.nn.Flatten(),
+            torch.nn.Linear(686, 5),
+        ),
+        "loss_function_fn": lambda: torch.nn.CrossEntropyLoss(reduction="mean"),
+        "target_fn": lambda: classification_targets((2,), 5),
+    },
+]
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/diaggnn_settings.py:239]: Can you give me an update how these settings differ from the automated simple-cnn tests we have set up?

File: backpack/extensions/secondorder/diag_ggn/activations.py [hunk 6]
```diff
@@ -17,3 +21,23 @@ def __init__(self):
...
+
+
+class DiagGGNLeakyReLU(DiagGGNBaseModule):
+    def __init__(self):
+        super().__init__(derivatives=LeakyReLUDerivatives())
+
+
+class DiagGGNLogSigmoid(DiagGGNBaseModule):
+    def __init__(self):
+        super().__init__(derivatives=LogSigmoidDerivatives())
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/activations.py:43]: Do we have tests for these? I'm just confused because the PR says only `ConvNd` and `ConvTransposeNd`are extended

Comment (sbharadwajj) [reply backpack/extensions/secondorder/diag_ggn/activations.py:43]: Umm, actually we dont have it here yet. We have it for `diag-h`

File: backpack/utils/conv.py (hunk 4 above)
Comment (sbharadwajj) [reply backpack/utils/conv.py:57]: Umm, we can generalize it. It's used in two places but the readability of the code will reduce

File: test/extensions/secondorder/diag_ggn/diaggnn_settings.py (hunk 5 above)
Comment (sbharadwajj) [reply test/extensions/secondorder/diag_ggn/diaggnn_settings.py:239]: Yes! So each of them differs in the configuration of `torch.nn.ConvTransposed(...)` as in, they have different values of padding, dilation, stride.. etc
I tried to make it automated but that just requires too many additional parameters whereas previous in activations we had only `bias=True/False`

File: backpack/extensions/secondorder/diag_ggn/activations.py (hunk 6 above)
Comment (sbharadwajj) [reply backpack/extensions/secondorder/diag_ggn/activations.py:43]: Added in local settings

File: backpack/utils/conv.py (hunk 4 above)
Comment (f-dangel) [reply backpack/utils/conv.py:57]: Wouldn't it also be used by transpose convolution>

File: backpack/utils/conv_transpose.py [hunk 7]
```diff
@@ -23,6 +24,55 @@ def get_weight_gradient_factors(input, grad_out, module, N):
...
+    input_reshaped = input.reshape(M, -1, spatial_out_numel)
+
+    if N == 1:
+        grad_output_viewed = grad_output
+    elif N == 2:
+        grad_output_viewed = eingroup("v,n,c,h,w->v,n,c,hw", grad_output)
+    elif N == 3:
+        grad_output_viewed = eingroup("v,n,c,d,h,w->v,n,c,dhw", grad_output)
+    else:
+        raise ValueError("{}-dimensional Conv. is not implemented.".format(N))
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:51]: transpose

File: backpack/utils/conv_transpose.py [hunk 8]
```diff
@@ -23,6 +24,55 @@ def get_weight_gradient_factors(input, grad_out, module, N):
...
+    V_axis, N_axis = 0, 1
+
+    if N == 1:
+        einsum_eq = "vncl->vnc"
+    elif N == 2:
+        einsum_eq = "vnchw->vnc"
+    elif N == 3:
+        einsum_eq = "vncdhw->vnc"
+    else:
+        ValueError("{}-dimensional Conv. is not implemented.".format(N))
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:72]: transpose

File: backpack/utils/conv.py (hunk 4 above)
Comment (sbharadwajj) [reply backpack/utils/conv.py:57]: yes!

Comment (sbharadwajj) [reply backpack/utils/conv.py:57]: okay when I change this, `test_kfac_shape[Conv2d-ReLU-regression-cpu]` this fails

Comment (f-dangel) [reply backpack/utils/conv.py:57]: Then, let's postpone this refactoring step until we have more extensive tests on KFAC and actually start extending it.
